### ISSUE TITLE
null ID should not be considered cacheable keys

### ISSUE THREAD

**Issue by mschipperheyn** [apollographql/apollo-client#9814]
**Intended outcome:**
I have a server side response that contains an array with two different objects with null IDs. I would expect that the missing cache key would not be interpreted as an actual cache key and I would get 2 different objects on the client side. 

**Actual outcome:**
What I'm getting is an array with 2 identical objects. A missing cache key should be considered as a "unique value".

**How to reproduce the issue:**
Graphql
```
type Product {
	id: ID!
	label: String!
	sku: String!
	productType: ProductType!
	costCategory: CostCategory!
	assetRangeFrom: Float
	assetRangeTo: Float
	state: State
	city: String
	year: Int
	costType: CostType!
	price: Float
	pricePercentage: Float
	costFloor: Float
	costCeiling: Float
	costExemption: Float
	charge: Boolean!
	createdAt: DateTime!
	updatedAt: DateTime!
}
type OrderLine {
	id: ID
	quantity: Int!
	product: Product!
	percentageValueBase: Float
	createdAt: DateTime
	updatedAt: DateTime
}
```

server side query results (Note the different id)
```
[
  {
    productId: '42df9f30-db9e-11ec-aaef-f158a82a932b',
    product: Product {
      dataValues: [Object],
      _previousDataValues: [Object],
      uniqno: 1,
      _changed: [Set],
      _options: [Object],
      isNewRecord: false,
      cache: [Function: cache],
      id: [Getter/Setter],
      searchText: [Getter/Setter],
      label: [Getter/Setter],
      sku: [Getter/Setter],
      productType: [Getter/Setter],
      costCategory: [Getter/Setter],
      costType: [Getter/Setter],
      state: [Getter/Setter],
      year: [Getter/Setter],
      assetRangeFrom: [Getter/Setter],
      assetRangeTo: [Getter/Setter],
      price: [Getter/Setter],
      pricePercentage: [Getter/Setter],
      costFloor: [Getter/Setter],
      costCeiling: [Getter/Setter],
      costExemption: [Getter/Setter],
      charge: [Getter/Setter],
      primaryKeyAttributes: [Array]
    },
    quantity: 1,
    percentageValueBase: 10000
  },
  {
    productId: 'd68daa10-daef-11ec-bfb6-55f481ef4749',
    product: Product {
      dataValues: [Object],
      _previousDataValues: [Object],
      uniqno: 1,
      _changed: [Set],
      _options: [Object],
      isNewRecord: false,
      cache: [Function: cache],
      id: [Getter/Setter],
      searchText: [Getter/Setter],
      label: [Getter/Setter],
      sku: [Getter/Setter],
      productType: [Getter/Setter],
      costCategory: [Getter/Setter],
      costType: [Getter/Setter],
      state: [Getter/Setter],
      year: [Getter/Setter],
      assetRangeFrom: [Getter/Setter],
      assetRangeTo: [Getter/Setter],
      price: [Getter/Setter],
      pricePercentage: [Getter/Setter],
      costFloor: [Getter/Setter],
      costCeiling: [Getter/Setter],
      costExemption: [Getter/Setter],
      charge: [Getter/Setter],
      primaryKeyAttributes: [Array]
    },
    quantity: 1,
    percentageValueBase: null
  }
]
```

client side query result (note the identical ID)
```
[
  {
    __typename: 'OrderLine',
    id: null,
    quantity: 1,
    percentageValueBase: 10000,
    product: {
      __typename: 'Product',
      id: '42df9f30-db9e-11ec-aaef-f158a82a932b',
      label: 'Taxa de serviço advogado',
      sku: 'ASSETFEE_2022_SP',
      productType: 'ASSET',
      costCategory: 'LAWYER',
      assetRangeFrom: null,
      assetRangeTo: null,
      state: 'SP',
      city: null,
      year: 2022,
      costType: 'PERCENTAGE',
      price: null,
      pricePercentage: 0.09,
      costExemption: 0,
      costFloor: 965,
      costCeiling: null,
      charge: false,
      updatedAt: '2022-05-24T20:15:31.000Z',
      createdAt: '2022-05-24T20:15:31.000Z'
    }
  },
  {
    __typename: 'OrderLine',
    id: null,
    quantity: 1,
    percentageValueBase: 10000,
    product: {
      __typename: 'Product',
      id: '42df9f30-db9e-11ec-aaef-f158a82a932b',
      label: 'Taxa de serviço advogado',
      sku: 'ASSETFEE_2022_SP',
      productType: 'ASSET',
      costCategory: 'LAWYER',
      assetRangeFrom: null,
      assetRangeTo: null,
      state: 'SP',
      city: null,
      year: 2022,
      costType: 'PERCENTAGE',
      price: null,
      pricePercentage: 0.09,
      costExemption: 0,
      costFloor: 965,
      costCeiling: null,
      charge: false,
      updatedAt: '2022-05-24T20:15:31.000Z',
      createdAt: '2022-05-24T20:15:31.000Z'
    }
  }
]
```

**Versions**
```
 System:
    OS: macOS 12.2.1
  Binaries:
    Node: 16.15.0 - ~/.nvm/versions/node/v16.15.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 8.5.5 - ~/.nvm/versions/node/v16.15.0/bin/npm
  Browsers:
    Chrome: 102.0.5005.61
    Safari: 15.3
  npmPackages:
    @apollo/client: ^3.6.7 => 3.6.7 
```

**Comment by mschipperheyn** [comment]
I solved this by using a typePolicy referencing the product.id, which in this case is unique for each query result. But I don't think this should be necessary. The default behaviour is counter intuitive and can easily lead to unexpected errors.

**Comment by mschipperheyn** [comment]
Another approach would be to exclude the id for let's say non id scenario's. But it all seems a bit of overkill. null ids should simply not be considered cacheable keys

**Comment by benjamn** [comment]
@mschipperheyn Sorry to keep you waiting here, but I fully agree the default behavior should be to treat multiple objects with `id: null` as unrelated, rather than treating them as the same object.

**Comment by bignimbus** [comment]
Hi all 👋🏻  I'm doing some housekeeping and noticed that this issue should have been closed by #9862.  So I'm closing the loop and therefore, this issue!  Let us know if you need more support here 🙏🏻
